### PR TITLE
Handle provider download/config

### LINKED ISSUE #15: Deployment and First run experience
Build package and publish as docker image. 

How is the provider shipped? 

1. Look at the TF registery APIs and download provider by name
2. Point to file (user could configure volume mount with provider)

In both instances pinning to a version of the provider is key - CRD needed to track the provider version used as after CRD creation provider cannot be changed.

### PR DESCRIPTION
**Pull request by lawrencegripper**
You can configure which terraform provider is used by the operator in two ways. The provider is automatically downloaded if `TF_PROVIDER_NAME` and `TF_PROVIDER_NAME` are set or is pulled from an existing location on disk if `TF_PROVIDER_NAME` and `TF_PROVIDER_NAME` are set.

- `TF_PROVIDER_NAME`    - Terraform provider you'd like to use. eg `azurerm`
- `TF_PROVIDER_VERSION` - Version of the terraform provider. eg `2.20.0`
- `TF_PROVIDER_PATH`    - Path on disk to the binary of the provider. eg "/providers/here". This folder must contain a provider binary in the format `terraform-provider-azurerm_v2.22`

The PR uses the following experimental Hashicorp libs to do this:

"github.com/hashicorp/terraform-exec/tfexec"
"github.com/hashicorp/terraform-exec/tfinstall"

Closes #15

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -24,6 +24,8 @@ terraform-hack-init:
 
 integration-tests: run
 	./scripts/wait-for-bridge.sh
+	@echo "==> Attempting to sourcing .env file"
+	if [ -f .env ]; then set -o allexport; . ./.env; set +o allexport; fi; \
```
Comment (jjcollinge) [inline Makefile:28]: There's gotta be a way we can export this across all sub shells - could try the top answer from here: https://stackoverflow.com/questions/44628206/how-to-load-and-export-variables-from-an-env-file-in-makefile?

**Comment by lawrencegripper** [comment]
Agreed there is likely a nicer way to do this. 

I'm rolling back the ginkgo conversion for the tests involved here as they're WAY more complex imo with ginkgo than just plain go tests. Makes sense for more complex tests but for this simple ones the standard go tests seem like a nicer fit.

**Comment by lawrencegripper** [comment]
Other than the makefile Env bit which I'll pickup - think this is good for a review now.

File: tfprovider/provider_test.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+			// Set required envs
+			for name, value := range tt.requiredEnvs {
+				err = os.Setenv(name, value)
+				if err != nil {
+					panic(err)
+				}
+			}
+			provider, err := getInstanceOfProvider(tt.name, providerInstallPath)
+			if err != nil {
+				t.Errorf("failed to get instance of provider. error = %v", err)
```
Comment (jjcollinge) [inline tfprovider/provider_test.go:63]: what's the best way to handle returned errors here? Seems we have panic, just log and log and return

File: Makefile (hunk 1 above)
Comment (lawrencegripper) [reply Makefile:30]: So turns out it's even easier than that. https://unix.stackexchange.com/a/348432

```
# Load the environment variables. If this errors review the README.MD and create a .env file as instructed
include .env
export
```

Comment (lawrencegripper) [reply Makefile]: FYI @EliiseS this seems to work well as long as you don't `"` variables, might be useful for future stuff. Also let me know if this is something you ruled out and I'm missing a flaw with the approach.

Comment (EliiseS) [reply Makefile]: I tried that, but it ended up not working with the "feature {}", so I went with the hacky approach.

Comment (lawrencegripper) [reply Makefile]: Hmmm wonder if this varies between machines. For me with 

This `Makefile`
```makefile
include .env
export

testenv:
	echo ${PROVIDER_CONFIG_HCL}
	$(shell echo $$PROVIDER_CONFIG_HCL)
```
with this `.env`
```bash
PROVIDER_CONFIG_HCL=features {}
TF_PROVIDER_NAME=azurerm
TF_PROVIDER_PATH=./hack/.terraform/plugins/linux_amd64/
```

outputs:

image

Double checked and the env isn't in the current bash shell already:

image

**Comment by lawrencegripper** [comment]
So TF 0.13 just released and broke this - now pinned to '0.12.29' and working again. Need to revisit and support 0.13 - will track as new issue

File: Makefile [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
 	go build .
 
 run: kind-create terraform-hack-init
-	@echo "==> Attempting to sourcing .env file"
-	if [ -f .env ]; then set -o allexport; . ./.env; set +o allexport; fi; \
-	go run . &
+	# Stop any previously running instance of the operator
+	$(shell [ -f run.pid ] && cat run.pid | xargs kill)
+	# Store the pid of the running instance in run.pid file
+	go run . & echo "$$$$" > run.pid
```
Comment (lawrencegripper) [inline Makefile:14]: @jjcollinge we talked about cleaning up the instance of the provider so `make integration-tests` can be re-run without things going badly. This does that by storing the PID of the instance in `run.pid` then cleaning up if the file exists before starting the provider.

Comment (lawrencegripper) [reply Makefile:14]: Seem like a reasonable solution?

Comment (jjcollinge) [reply Makefile:14]: Cool looks good - maybe flock the pid file to ensure only one run. Maybe we revisit this if we orchestrate the integrations tests from Go eventually?

Comment (lawrencegripper) [reply Makefile:14]: Sorry, totally forgot we talked about that - yeah that is a much better approach. Deal lets keep this for now and I'll raise an issue to track moving to go orchestrating the operators start up/shutdown.

**Comment by lawrencegripper** [comment]
Rebased onto `main` once build goes green I'll merge.

**Pull request merged**